include/deal.II/cgal/triangulation.h

```diff
@@ -0,0 +1,259 @@
+  template <typename CGALTriangulation, int dim, int spacedim>
+  void
+  cgal_triangulation_to_dealii_triangulation(
+    const CGALTriangulation &     cgal_triangulation,
+    Triangulation<dim, spacedim> &dealii_triangulation);
+
+
+
+#  ifndef DOXYGEN
+  // Template implementation
+  template <int spacedim, typename CGALTriangulation>
+  void
+  add_points_to_cgal_triangulation(const std::vector<Point<spacedim>> &points,
+                                   CGALTriangulation &triangulation)
+  {
+    Assert(triangulation.is_valid(),
+           ExcMessage(
+             "The triangulation you pass to this function should be a valid "
+             "CGAL triangulation."));
+    using CGALPoint = typename CGALTriangulation::Point;
+    std::vector<CGALPoint> cgal_points(points.size());
+    std::transform(
+      points.begin(), points.end(), cgal_points.begin(), [](const auto &p) {
+        return CGALWrappers::dealii_point_to_cgal_point<CGALPoint>(p);
+      });
+
+    triangulation.insert(cgal_points.begin(), cgal_points.end());
+    Assert(triangulation.is_valid(),
+           ExcMessage(
+             "The Triangulation is no longer valid after inserting the points. "
+             "Bailing out."));
+  }
+
+
+
+  template <typename CGALTriangulation, int dim, int spacedim>
+  void
+  cgal_triangulation_to_dealii_triangulation(
+    const CGALTriangulation &     cgal_triangulation,
+    Triangulation<dim, spacedim> &dealii_triangulation)
+  {
+    AssertThrow(cgal_triangulation.dimension() == dim,
+                ExcMessage("The dimension of the input CGAL triangulation (" +
+                           std::to_string(cgal_triangulation.dimension()) +
+                           ") does not match the dimension of the output "
+                           "deal.II triangulation (" +
+                           std::to_string(dim) + ")."));
+
+    Assert(dealii_triangulation.n_cells() == 0,
+           ExcMessage("The output triangulation object needs to be empty."));
+
+    const auto nv = cgal_triangulation.number_of_vertices();
+
+    // Deal.II storage data structures
+    std::vector<Point<spacedim>> vertices(nv);
+    std::vector<CellData<dim>>   cells;
+    SubCellData                  subcell_data;
+
+    // CGAL storage data structures
+    std::map<typename CGALTriangulation::Vertex_handle, unsigned int>
+      vertex_map;
+    {
+      unsigned int i = 0;
+      for (auto v : cgal_triangulation.finite_vertex_handles())
+        {
+          vertices[i] =
+            CGALWrappers::cgal_point_to_dealii_point<spacedim>(v->point());
+          vertex_map[v] = i++;
+        }
+    }
+
+    auto has_faces = boost::hana::is_valid(
+      [](auto &&obj) -> decltype(obj.finite_face_handles()) {});
+
+    auto has_cells = boost::hana::is_valid(
+      [](auto &&obj) -> decltype(obj.finite_cell_handles()) {});
+
+    // Different loops for Triangulation_2 and Triangulation_3 types.
+    if constexpr (decltype(has_faces(cgal_triangulation)){})
```

Comment:
We cannot use `if constexp` yet.  That is only available once we switch to C++17. You will need to do template specialization.

Reply:
CGALWrappers are enabled only if c++17 is enabled.

Reply:
OK. I see.